src/com/tony/server/ResponseDeterminer.java

```diff
@@ -23,12 +30,37 @@ public Response determineResponse(Request request) {
             if(authenticator.requiresAuthentication(request)
                     && !authenticator.isAuthenticated(request)){
                 return new UnauthorizedResponse();
+            } else if(request.getHttpMethod().equals("PATCH")){
+                String filePath = ((PatchResponse)router.route(request)).getFilePath();
+                if(!fileEtagMatchesGivenEtag(filePath, request.getHeaders().get("If-Match"))) {
+                    return new PreconditionFailedResponse();
+                } else {
+                    return router.route(request);
+                }
             } else {
                 return router.route(request);
             }
         } else if(uriList.contains(request.getURI())){
             return new MethodNotAllowedResponse();
+        } else if(request.getURI().startsWith("/parameters?")){
+            return new ParameterDecodeResponse(request.getURI());
         }
         return new FourOhFourResponse();
     }
```

Comment:
refactored in https://github.com/tbaik/java-server/commit/d07ddfc79a621f584cf14125a0d4127e1edcef64